### PR TITLE
Add test for devices interaction with default flow

### PR DESCRIPTION
**Pull request by rtretyak**

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowConnectedDeviceSpec.groovy [hunk 1]
```diff
@@ -637,16 +658,50 @@ srcLldpDevices=#newSrcEnabled, dstLldpDevices=#newDstEnabled"() {
...
         Wrappers.wait(WAIT_OFFSET) {
-            with(northbound.getFlowConnectedDevices(flow.id)) {
-                it.source.lldp.size() == 1
-                verifyEquals(it.source.lldp.first(), lldpData)
+            with(northboundV2.getConnectedDevices(sw.dpId).ports) {
+                it.size() == 1
+                it[0].portNumber == tg.switchPort
+                it[0].lldp.size() == 2
+                it[0].lldp.last().flowId == defaultFlow.id
+                it[0].lldp.last().vlan == vlan
```
Comment (niksv) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowConnectedDeviceSpec.groovy]: 1. the order of devices is not guarantee, so the device can be the first in some cases.  
2. There is one limitation for default flows - vlan for coming packets equals to 0. Even if you push packet with vlan 11 to default flow - you will get connected device with vlan 0.

Could you please add more tests for default rules? For example:
one switch flow:
inVlan: 0,  outVlan: 0
SeveralSwitchFlow:
inVlan: 0,  outVlan: 0
inVlan: 0,  outVlan: X

Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowConnectedDeviceSpec.groovy]: 1. Do you find it reasonable to sort devices by timeLastSeen? I would expect some ordering there

Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowConnectedDeviceSpec.groovy]: 2. So to clarify - I won't be able to see vlans neither per-switch nor per-flow if there is a default flow present?

Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowConnectedDeviceSpec.groovy]: Added 
```
SeveralSwitchFlow:
inVlan: 0, outVlan: 0
inVlan: X, outVlan: 0
```
Other cases tested manually.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy [hunk 2]
```diff
@@ -705,7 +886,9 @@ srcLldpDevices=#newSrcEnabled, dstLldpDevices=#newDstEnabled"() {
     private void enableMultiTableIfNeeded(boolean lldpEnabled, SwitchId switchId) {
```
Comment (andriidovhan) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: what about adding additional parameters (e.g initSwProps) ?
then you can just do:
```
!initSwProps.multiTable && northbound.updateSwitchProperties(sw.dpId, initSwProps.jacksonCopy().tap {
            it.multiTable = true
        })
```
What I mean is you can avoid sending needless get request `northbound.getSwitchProperties(sw.dpId)`

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy [hunk 3]
```diff
@@ -637,21 +669,144 @@ srcLldpDevices=#newSrcEnabled, dstLldpDevices=#newDstEnabled"() {
...
+
+        where:
+        srcDefault | dstDefault
+        false      | true
+        true       | true
+
+        descr = "default(no-vlan) flow endpoints on ${getDescr(srcDefault, dstDefault)}"
     }
 
     def "System forbids to turn on 'connected devices per switch' on a single-table-mode switch"() {
```
Comment (andriidovhan) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: Does system allow to disable `multiTableMode` when `lldp` is enabled?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy [hunk 4]
```diff
@@ -637,21 +669,144 @@ srcLldpDevices=#newSrcEnabled, dstLldpDevices=#newDstEnabled"() {
...
         Wrappers.wait(WAIT_OFFSET) {
-            with(northbound.getFlowConnectedDevices(flow.id)) {
-                it.source.lldp.size() == 1
-                verifyEquals(it.source.lldp.first(), lldpData)
+            with(northboundV2.getConnectedDevices(sw.dpId).ports) {
+                it.size() == 1
+                it[0].portNumber == tg.switchPort
+                it[0].lldp.size() == 2
+                it[0].lldp.last().flowId == defaultFlow.id
+                //this behavior explained in 'where' block comment
```
Comment (andriidovhan) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: I don't see any comment in the `where`  block

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy [hunk 5]
```diff
@@ -877,4 +1069,16 @@ srcLldpDevices=#newSrcEnabled, dstLldpDevices=#newDstEnabled"() {
...
+        if (src && !dst) {
+            "src only"
+        } else if (!src && dst) {
+            "dst only"
+        } else if (src && dst) {
+            "src and dst"
+        } else {
+            "none of the endpoints"
+        }
+    }
```
Comment (andriidovhan) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: We have to ignore all tests where the `multiTable` feature is required until #2971 is fixed.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy [hunk 6]
```diff
@@ -510,11 +521,18 @@ srcLldpDevices=#newSrcEnabled, dstLldpDevices=#newDstEnabled"() {
...
+        with(northboundV2.getConnectedDevices(sw.dpId).ports) {
+            it.size() == 1
+            it[0].portNumber == tg.switchPort
+            it[0].lldp.first().vlan == flow.source.vlanId
+            it[0].lldp.first().flowId == flow.id
+            verifyEquals(it[0].lldp.first(), lldpData)
+        }
 
         cleanup: "Remove created flow and device"
         flow && northbound.deleteFlow(flow.id)
```
Comment (niksv) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: do we need `flowHelperV2.deleteFlow(flow.id)` here and in line 291?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy [hunk 7]
```diff
@@ -558,7 +577,10 @@ srcLldpDevices=#newSrcEnabled, dstLldpDevices=#newDstEnabled"() {
...
 
         then: "Device is registered as a flow device"
-        !northbound.getFlowConnectedDevices(flow.id).source.lldp.empty
+        with(northbound.getFlowConnectedDevices(flow.id)) {
+            it.source.lldp.size() == 1
+            verifyEquals(it.source.lldp.first(), lldpData)
+        }
 
         cleanup: "Remove created flow and registered devices, revert switch props"
         northbound.deleteFlow(flow.id)
```
Comment (niksv) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: do we need `flowHelperV2.deleteFlow(flow.id)` here?

Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: yeah we do, but there might be a defect with v2 delete there. Lemme check that tomorrow

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy [hunk 8]
```diff
@@ -637,21 +669,144 @@ srcLldpDevices=#newSrcEnabled, dstLldpDevices=#newDstEnabled"() {
...
+        where:
+        srcDefault | dstDefault
+        true       | false
+        true       | true
+
+        descr = "default ${getDescr(srcDefault, dstDefault)}"
+    }
+
+    @Unroll
+    def "System properly detect device vlan in case of #descr"() {
```
[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: done

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy (hunk 2 above)
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: done

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy (hunk 4 above)
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: removed the comment. It used to describe some weird behavior here which no longer exists

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy (hunk 5 above)
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: we are deleting via v1 instead. Is this an option?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy (hunk 3 above)
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ConnectedDevicesSpec.groovy]: Nice catch, let me try

**Pull request merged**